lib/Dialect/Handshake/HandshakeOps.cpp

```diff
@@ -2018,5 +2018,23 @@ LogicalResult TruncFOp::verify() { return verifyTruncOp(*this); }
 //===----------------------------------------------------------------------===//
 LogicalResult ExtFOp::verify() { return verifyExtOp(*this); }
 
+//===----------------------------------------------------------------------===//
+// ExtractOp
+//===----------------------------------------------------------------------===//
+LogicalResult ExtractOp::inferReturnTypes(
+    MLIRContext *context, std::optional<Location> location, ValueRange operands,
+    DictionaryAttr attributes, OpaqueProperties, RegionRange,
+    SmallVectorImpl<Type> &inferredReturnTypes) {
+
+  Type channelLikeType = operands.front().getType();
+  if (auto channelType = dyn_cast<handshake::ChannelType>(channelLikeType)) {
+    inferredReturnTypes.push_back(ChannelType::get(channelType.getDataType()));
+  } else {
+    assert(isa<handshake::ControlType>(channelLikeType) && "expected control");
+    inferredReturnTypes.push_back(handshake::ControlType::get(context));
+  }
```

Comment:
How about:

```cpp
auto channelType = cast<handshake::ExtraSignalsTypeInterface>(operands.front().getType());
inferredReturnTypes.push_back(channelType.copyWithExtraSignals({}));
return success();
```

- Both `ChannelType` and `ControlType` implement `ExtraSignalsTypeInterface`.
- `copyWithExtraSignals({})` creates a new type with the empty extra signals, but keeps the data type if present.
  - We can drop this weird interface after unifying `ChannelType` and `ControlType` (#336).
- The `cast` is safe: it's an LLVM `cast`, which asserts on failure. I think we don't need if statement here since type verification guarantees this cast succeeds (the same approach is taken in `ConstantOp::inferReturnTypes` and `CmpIOp::inferReturnTypes`).